.teamcity/src/main/kotlin/configurations/Gradleception.kt

```diff
@@ -20,23 +26,61 @@ class Gradleception(model: CIBuildModel, stage: Stage) : BaseGradleBuildType(sta
         javaCrash = false
     }
 
+    params {
+        // Override the default commit id so the build steps produce reproducible distribution
+        param("env.BUILD_COMMIT_ID", "HEAD")
+    }
+
+    /*
+     To avoid unnecessary rerun, what we do here is a bit complicated:
+
+     1. Build a Gradle distribution with a fixed timestamp and hash it, but never use this distribution.
+     2. Build a Gradle distribution with this hash as a version + fixed timestamp -> dogfood-first
+     3. Build a Gradle distribution using dogfood-first with the "hash of dogfood-first" as a version + fixed timestamp different from the one above -> dogfood-second
+     4. Use dogfood-second to run a Gradle build which runs tests.
+
+     */
+    val dogfoodTimestamp1 = "19800101010101+0000"
+    val dogfoodTimestamp2 = "19800202020202+0000"
     val buildScanTagForType = buildScanTag("Gradleception")
     val defaultParameters = (buildToolGradleParameters() + listOf(buildScanTagForType) + "-Porg.gradle.java.installations.auto-download=false").joinToString(separator = " ")
 
-    applyDefaults(model, this, ":distributions-full:install", notQuick = true, extraParameters = "-Pgradle_installPath=dogfood-first $buildScanTagForType", extraSteps = {
-        localGradle {
-            name = "BUILD_WITH_BUILT_GRADLE"
-            tasks = "clean :distributions-full:install"
-            gradleHome = "%teamcity.build.checkoutDir%/dogfood-first"
-            gradleParams = "-Pgradle_installPath=dogfood-second -PignoreIncomingBuildReceipt=true $defaultParameters"
-        }
-        localGradle {
-            name = "QUICKCHECK_WITH_GRADLE_BUILT_BY_GRADLE"
-            tasks = "clean sanityCheck test"
-            gradleHome = "%teamcity.build.checkoutDir%/dogfood-second"
-            gradleParams = defaultParameters
-        }
-    })
+    applyDefaults(
+        model,
+        this,
+        ":distributions-full:install",
+        notQuick = true,
+        extraParameters = "-Pgradle_installPath=dogfood-first-for-hash -PignoreIncomingBuildReceipt=true -PbuildTimestamp=$dogfoodTimestamp1 $buildScanTagForType",
+        extraSteps = {
+            script {
+                name = "CALCULATE_MD5_VERSION_FOR_DOGFOODING_DISTRIBUTION"
+                workingDir = "%teamcity.build.checkoutDir%/dogfood-first-for-hash"
+                scriptContent = """
+                    set -x
+                    MD5=`find . -type f | sort | xargs md5sum | md5sum | awk '{ print ${'$'}1 }'`
```

Comment:
```suggestion
                    MD5=`find . -type f | sort | xargs md5sum | md5sum | awk '{ print $1 }'`
```

I don't understand why you need to escape `$` here, given that it is inclosed in single ticks `'`. The given command doesn't work in my command-line. Is Teamcity changing something here?

Reply:
The `$` escaping here has nothing to do with single quote `'`, but because it's in Kotlin double quotes. Though it can be remove because `$1` is not a valid kotlin expression. So either `find . -type f | sort | xargs md5sum | md5sum | awk '{ print ${'$'}1 }'` and `find . -type f | sort | xargs md5sum | md5sum | awk '{ print $1 }'` will be rendered as same `find . -type f | sort | xargs md5sum | md5sum | awk '{ print $1 }'`.

What doesn't work for you?

```
$ MD5=`find . -type f | sort | xargs md5sum | md5sum | awk '{ print $1 }'`

# zhb @ zhb in ~/Projects/gradle/gradle on git:blindpirate/faster-gradleception x [17:41:51]
$ echo $MD5
cec46cb6c1bf9297aa7c9ad8080f602b

```

Reply:
Ah, got it. Using the `${'$'}` in the shell directly didn't work.